Fallback to single episode search for found form Trakt errors

glensc opened the pull request:
Replaces https://github.com/Taxel/PlexTraktSync/pull/240, which proven to be too slow to fetch from trakt for each episode.

So this PR uses fetch for episode only if previous bulk fetch failed.

- [x] Solve `WARNING: Skipping <imdb:tt11909222:<Episode:198150:Frank-of-Ireland-s01e01>>: Not found on Trakt`
- [ ] Solve repeated `INFO: Add to Trakt Collection: <imdb:tt11909222:<Episode:198150:Frank-of-Ireland-s01e01>>`

Problem solution to the discussions at:
- https://github.com/Taxel/PlexTraktSync/discussions/208
- https://github.com/Taxel/PlexTraktSync/pull/240

glensc commented:
@BEisem: Can you confirm the first problem listed in the PR body is solved and that the performance of the full scan is acceptable.

BEisem commented:
Looks like this initial scan will take awhile, but it's definitely reasonable for a first scan. Much better than before when it was deep scanning everything.

As for fixing the problem ... it is still repeatedly reporting "Add to Trakt Collection".

```
bink@Homey-McPlexFace plex-trakt-sync-episode-lookup-fallback % ./plex_trakt_sync.sh sync --show="Frank of Ireland"
INFO: Syncing with Plex BlaEis and Trakt binkatl
INFO: Syncing Show: Frank of Ireland
INFO: Loaded Trakt lists in 3.5 seconds
INFO: Server version 1.22.2.4282-a97b03fad updated at: 2021-04-16 14:01:05
INFO: Recently added: [<Movie:123435:Billy-Connolly:-Live>, <Movie:198518:Psycho>, <Movie:198515:Pirates-of-the-Carib>, <Movie:198499:Philomena>, <Movie:198479:Pulp-Fiction>]
INFO: Add to Trakt Collection: <imdb:tt11909222:<Episode:198150:Frank-of-Ireland-s01e01>>
INFO: Add to Trakt Collection: <imdb:tt11909226:<Episode:198167:Frank-of-Ireland-s01e02>>
INFO: Add to Trakt Collection: <imdb:tt11909228:<Episode:198168:Frank-of-Ireland-s01e03>>
INFO: Add to Trakt Collection: <imdb:tt11909232:<Episode:198169:Frank-of-Ireland-s01e04>>
INFO: Add to Trakt Collection: <imdb:tt11909234:<Episode:198170:Frank-of-Ireland-s01e05>>
INFO: Add to Trakt Collection: <imdb:tt11909236:<Episode:198171:Frank-of-Ireland-s01e06>>
INFO: Updated plex watchlist in 0.0 seconds
INFO: Updated Trakt collection: {'existing': {'episodes': 6}}
INFO: Completed full sync in 7.3 seconds
bink@Homey-McPlexFace plex-trakt-sync-episode-lookup-fallback % ./plex_trakt_sync.sh sync --show="Frank of Ireland"
INFO: Syncing with Plex BlaEis and Trakt binkatl
INFO: Syncing Show: Frank of Ireland
INFO: Loaded Trakt lists in 3.5 seconds
INFO: Server version 1.22.2.4282-a97b03fad updated at: 2021-04-16 14:01:05
INFO: Recently added: [<Movie:123435:Billy-Connolly:-Live>, <Movie:198518:Psycho>, <Movie:198515:Pirates-of-the-Carib>, <Movie:198499:Philomena>, <Movie:198479:Pulp-Fiction>]
INFO: Add to Trakt Collection: <imdb:tt11909222:<Episode:198150:Frank-of-Ireland-s01e01>>
INFO: Add to Trakt Collection: <imdb:tt11909226:<Episode:198167:Frank-of-Ireland-s01e02>>
INFO: Add to Trakt Collection: <imdb:tt11909228:<Episode:198168:Frank-of-Ireland-s01e03>>
INFO: Add to Trakt Collection: <imdb:tt11909232:<Episode:198169:Frank-of-Ireland-s01e04>>
INFO: Add to Trakt Collection: <imdb:tt11909234:<Episode:198170:Frank-of-Ireland-s01e05>>
INFO: Add to Trakt Collection: <imdb:tt11909236:<Episode:198171:Frank-of-Ireland-s01e06>>
INFO: Updated plex watchlist in 0.0 seconds
INFO: Updated Trakt collection: {'existing': {'episodes': 6}}
INFO: Completed full sync in 5.7 seconds
```

glensc commented:
> As for fixing the problem ... it is still repeatedly reporting "Add to Trakt Collection".

I did not ask to test that. It's a known problem, no solution yet. I've only asked to test the first task in the PR body.

glensc commented:
The slow initial scan, is it slower than 0.7.x? 0.6.x? 0.5.x? as the initial full scan without cache has always been slow.

BEisem commented:
> > As for fixing the problem ... it is still repeatedly reporting "Add to Trakt Collection".
> 
> I did not ask to test that. It's a known problem, no solution yet. I've only asked to test the first task in the PR body.

Sorry, you had two lines listed in the PR body, I thought this was supposed to fix both ... I didn't realize one was checked and the second was not checked (the check boxes are greyed out and almost invisible on my screen). So yes, this solves the first problem (the one that is checked) ... it no longer skips due to "Not found on Trakt."

BEisem commented:
> The slow initial scan, is it slower than 0.7.x? 0.6.x? 0.5.x? as the initial full scan without cache has always been slow.

I'll have to do some actual tests with a timer. It "feels" like it is not quite as fast as 0.7 (understandable because of the deep scan for episodes that fail the main scan), but I don't remember the older ones since I haven't used them in awhile.

glensc commented:
I marked the first one checked after seeing your pasted output. Looked like confirmation. So at the time of asking, it was not checked.

glensc commented:
0.7.x has progressbar, so can just compare with that...

glensc commented:
And yes, this PR is supposed to fix both, as no code yet written for the second problem, no point in testing it.

BEisem commented:
Ok, testing on my "Reality TV" library (stopping the scan after that library finishes).

0.7.19 - 6:15
This PR - 19:20 (estimated by progress bar)... I'll let it run through to the end to see what the actual time is.

Also, one thing to keep in mind is, the amount of time this version takes will vary depending on how many "problematic" episodes there are in the library. If there are no problematic episodes, the time should essentially be the same as in 0.7.19.

glensc commented:
Amount of problematic items can be identified by counting repeated `Add to Trakt Collection` messages :)

glensc commented:
sorry, what is 6:15? 00:06:15 or 06:15:00?

and you started comparing with both empty cache?

BEisem commented:
Yes, both started with empty cache.

0.7.19 - 6 minutes 15 seconds
This PR - 19 minutes 20 seconds (at first). However I just checked again, and the initial scan is still running ... the estimate is now up to 6 hours 18 minutes. It has only completed 39 out of 418 shows in this library.

BEisem commented:
Another check, and now it's estimating 9 hours 51 minutes to complete. Only 50 out of 418 are finished. I don't think this is going to be a feasible solution, unfortunately...

glensc commented:
Excluding the tests (I've removed the tests for now for PR cleanness), the only code is to fall back to slow mode. is that really the source of slowness? does the amount of Adding to collection match with this estimation?

There are some other releases before this branch, can you test these too, maybe the slowness was introduced in earlier pull requests.

I've added WARNING message to print when the fallback call is made. Maybe this helps.

Also, you sure you not getting slowed down due to rate limits?

BEisem commented:
I will test the other releases. Can you let me know exactly which ones to test?

I don't think it's a rate limit, because the 0.7.19 version is still quick.

BEisem commented:
Seems like the problem happens between 0.7.19 and 0.7.20. I'm not sure if this is what you meant, but 0.7.19 finishes an initial scan in 5-6 minutes for my Reality TV library. On 0.7.20, it has been running for several hours and it has only gotten through 87 out of 418 shows. And it is estimating 4 hours 39 minutes remaining...

glensc commented:
This is good, I can bisect the changes between the two releases:
- https://github.com/Taxel/PlexTraktSync/compare/0.7.19...0.7.20

and I think I know what's the drag. 

I'm pretty sure that it's about `lookup = self.lookup(tm)` being called in a loop. I mean I hope so, as then it's easy to fix.

glensc commented:
Btw, I was thinking, maybe you can send me your Plex database? Then I can set up Plex with that database. I don't need the media files, just the database. Upload it to https://wetransfer.com/, and mail it to me. to obtain my email, find any commit from this repo and append `.patch` to the URL. to contain the database, visit Plex Web -> Manage Server -> Settings -> Troubleshooting -> Download database.

BEisem commented:
> Btw, I was thinking, maybe you can send me your Plex database? Then I can set up Plex with that database.

Sure, that's a good idea. I'm uploading the file now. Please let me know you got it...

glensc commented:
I think this fixes the performance: 808e186, this is based on your previous assessment that the problem was in those versions: https://github.com/Taxel/PlexTraktSync/pull/252#issuecomment-821747303

however, I can't feel much difference with muy TV library, processes in 25-35s with cache removed.

@BEisem with that applied (in this branch), you can confirm the performance is back to acceptable?

BEisem commented:
I'm definitely happy to test this and compare. Can you please confirm when you say "this branch" ... I'm still not sure how to determine which branch. Should I test it on https://github.com/glensc/plex-trakt-sync/tree/episode-lookup-fallback ?

glensc commented:
@BEisem yes, this branch = this pr = this pull request = https://github.com/Taxel/PlexTraktSync/pull/252 = glensc:episode-lookup-fallback

![image](https://user-images.githubusercontent.com/199095/115148761-be523180-a069-11eb-80e4-80839cc6b018.png)

btw, if you're that frequently switching branches, you may find `gh` tool useful, it's an answer to your question:
- https://github.com/Taxel/PlexTraktSync/discussions/236#discussioncomment-620516

however, it might be pandora's box of problems that are unknown to you. so maybe doesn't suit you. as zip download is still certainly most straightforward and reliable to get the job done.

BEisem commented:
Thank you. I thought that's what it meant, I just wanted to be 100% sure, if I'm testing the wrong one it doesn't help anyone so I just wanted to clarify.

I'm currently doing an initial sync (without an existing sqlite cache) and this seems to be very good. I think this would be very acceptable for a first scan. Thank you!

glensc commented:
This Frank-of-Ireland is weird. Searching by show id gives no results. So it's some kind of glitch on Trakt side:

- https://api.trakt.tv/shows/163427/seasons - empty
- https://api.trakt.tv/shows/game-of-thrones/seasons - it should work like this

the actual query is with `extended=episodes`, but that requires api keys;
- https://api.trakt.tv/shows/163427/seasons?extended=episodes
- https://api.trakt.tv/shows/game-of-thrones/seasons?extended=episodes

docs:
- https://trakt.docs.apiary.io/#reference/seasons/summary/get-all-seasons-for-a-show?console=1

BEisem commented:
I just completed an initial sync of my TV libraries in 46 minutes, 7 seconds. This is fantastic, as before it was estimating many hours to complete (I never completed a full scan at that speed).

glensc commented:
What do you think, is it worth to merge this with only first item being fixed? As fixing the second item, I have no idea yet how to fix it. May take time until I find solution.

And this PR discussion again has grown huge!

BEisem commented:
I think this is a great solution and definitely worth merging.

glensc commented:
Ok, merging the half baked variant then :D
